core/templates/dev/head/collection_editor/editor_tab/CollectionLinearizerServiceSpec.js

```diff
@@ -0,0 +1,468 @@
+                  thirdCollectionNode]);
+      expect(CollectionLinearizerService.removeCollectionNode(collection,
+          'exp_id1')).toBe(true);
+      expect(collection.containsCollectionNode('exp_id1')).toBe(false);
+      expect(
+          CollectionLinearizerService.getCollectionNodesInPlayableOrder(
+              collection)).toEqual([firstCollectionNode, thirdCollectionNode]);
+    });
+  });
+
+  describe('appendCollectionNode()', function() {
+    it('should correctly append a node to an empty collection', function() {
+      var collection = CollectionObjectFactory.createEmptyCollection();
+      expect(collection.containsCollectionNode('exp_id0')).toBe(false);
+      expect(
+          CollectionLinearizerService.getCollectionNodesInPlayableOrder(
+                  collection)).toEqual([]);
+      expect(firstCollectionNode.getAcquiredSkillList().getSkills()).toEqual(
+          []);
+      CollectionLinearizerService.appendCollectionNode(collection, 'exp_id0',
+          firstCollectionNode.getExplorationSummaryObject());
+      firstCollectionNode = collection.getCollectionNodeByExplorationId(
+          'exp_id0');
+      expect(
+          firstCollectionNode.getPrerequisiteSkillList().getSkills()).toEqual(
+              []);
+      expect(firstCollectionNode.getAcquiredSkillList().getSkills()).toEqual([
+          'exp title0']);
+      expect(
+          CollectionLinearizerService.getCollectionNodesInPlayableOrder(
+              collection)).toEqual([firstCollectionNode]);
+    });
+
+    it('should correctly append a node to a non-empty collection', function() {
+      var collection = createLinearCollection();
+      var newCollectionNodeBackendObject = {
+        exploration_id: 'exp_id3',
+        prerequisite_skills: [],
+        acquired_skills: [],
+        exploration_summary: {
+          title: 'exp title3',
+          category: 'exp category',
+          objective: 'exp objective'
+        }
+      };
+      var newCollectionNode = CollectionNodeObjectFactory.create(
+          newCollectionNodeBackendObject);
+      expect(newCollectionNode.getPrerequisiteSkillList().getSkills()).toEqual(
+          []);
+      expect(newCollectionNode.getAcquiredSkillList().getSkills()).toEqual(
+          []);
+      expect(
+          CollectionLinearizerService.getCollectionNodesInPlayableOrder(
+              collection)).toEqual([firstCollectionNode, secondCollectionNode,
+                  thirdCollectionNode]);
+      CollectionLinearizerService.appendCollectionNode(collection, 'exp_id3',
+          newCollectionNode.getExplorationSummaryObject());
+      newCollectionNode = collection.getCollectionNodeByExplorationId(
+        'exp_id3');
+      expect(newCollectionNode.getPrerequisiteSkillList().getSkills()).toEqual(
+          ['exp title2']);
+      expect(newCollectionNode.getAcquiredSkillList().getSkills()).toEqual(
+          ['exp title3']);
```

Comment:
Let's say we want to catch an issue where nodes are in playable order but there is a bug in the implementation where the skills are incorrectly updated. Where would the test for that be?